dttools/src/path_disk_size_info.c

```diff
@@ -47,102 +42,114 @@ int path_disk_size_info_get(const char *path, int64_t *measured_size, int64_t *n
 }
 
 int path_disk_size_info_get_r(const char *path, int64_t max_secs, struct path_disk_size_info **state) {
-	int64_t start_time = time(0);
-	int result = 0;
 
 	if(!*state) {
 		/* if state is null, there is no state, and path is the root of the measurement. */
 		*state = calloc(1, sizeof(struct path_disk_size_info));
 	}
-
 	struct path_disk_size_info *s = *state;     /* shortcut for *state, so we do not need to type (*state)->... */
 
+	int64_t start_time = time(0);
+	int result = 0;
+	
 	/* if no current_dirs, we begin a new measurement. */
 	if(!s->current_dirs) {
 		s->complete_measurement = 0;
+		s->current_dirs 		= list_create();
+		s->size_so_far 			= 0;
+		s->count_so_far 		= 0;
+		list_push_tail(s->current_dirs, (void*) xxstrdup(path));
+	}
 
-		struct DIR_with_name *here = malloc(sizeof(struct DIR_with_name));
+	/* Start processing unseen directories.
+	 * Each loop counts the current directory and all files immediately except 
+	 * files that are directories. Directory entries are added to a list to be 
+	 * counted later.
+	 */
+	char* tail_name;
+	struct dirent *entry;
+	struct stat file_info;
+	DIR* tail_dir;
+	int timeout = 0;
+	while((tail_name = list_pop_tail(s->current_dirs))) {
+		tail_dir = opendir(tail_name);
+		if (!tail_dir) {
+			if (errno == ENOENT) {
+				/* Directory might have been removed so we ignore this. */
+			}
+			else {
+				debug(D_DEBUG, "error opening directory '%s': (errno) %s\n", tail_name, strerror(errno));
+				result = -1;
+			}
+			free(tail_name);
+			continue;
+		}
 
-		if((here->dir = opendir(path))) {
-			here->name = xxstrdup(path);
-			s->current_dirs = list_create();
-			s->size_so_far  = 0;
-			s->count_so_far = 1;                     /* count the root directory */
-			list_push_tail(s->current_dirs, here);
-		} else {
-			debug(D_DEBUG, "error reading disk usage on directory: %s.\n", path);
-			s->size_so_far  = -1;
-			s->count_so_far = -1;
-			s->complete_measurement = 1;
-			result       = -1;
-
-			free(here);
-			goto timeout;
+		/* Add current dir to stats */
+		if(lstat(tail_name, &file_info) < 0) {
+			if(errno == ENOENT) {
+				/* Directory might have been removed so we ignore this. */
+			} else {
+				debug(D_DEBUG, "error reading disk usage on '%s'. errno: %s\n", path, strerror(errno));
+				result = -1;
+			}
+			closedir(tail_dir);
+			free(tail_name);
+			continue;
 		}
-	}
+		s->size_so_far += file_info.st_size;
+		s->count_so_far += 1;
 
-	struct DIR_with_name *tail;
-	while((tail = list_peek_tail(s->current_dirs))) {
-		struct dirent *entry;
-		struct stat   file_info;
-		while((entry = readdir(tail->dir))) {
+		/* Read entries of opened directory */
+		while((entry = readdir(tail_dir))) {
 			if( strcmp(".", entry->d_name) == 0 || strcmp("..", entry->d_name) == 0)
 				continue;
 
 			char composed_path[PATH_MAX];
```

Comment:
Can you explain how string_format can help that code?

Reply:
`char * path = string_format( ... )`

Then path is not limited to any particular length.  Also reduces stack consumption because you aren't allocating the maximum path length on every recursive function call.